client/pd_service_discovery.go

```diff
@@ -97,6 +100,237 @@ type ServiceDiscovery interface {
+	if c == nil {
+		return nil
+	}
+	return c.conn
+}
+
+// NeedRetry implements ServiceClient.
+func (c *pdServiceClient) NeedRetry(pdErr *pdpb.Error, err error) bool {
+	if c.IsLeader() {
+		return false
+	}
+	return !(err == nil && pdErr == nil)
+}
+
+type errFn func(pdErr *pdpb.Error) bool
+
+func regionAPIErrorFn(pdErr *pdpb.Error) bool {
+	return pdErr.GetType() == pdpb.ErrorType_REGION_NOT_FOUND
+}
+
+// pdServiceAPIClient is a specific API client for PD service.
+// It extends the pdServiceClient and adds additional fields for managing availability
+type pdServiceAPIClient struct {
+	ServiceClient
+	fn errFn
+
+	unavailable      atomic.Bool
+	unavailableUntil atomic.Value
+}
+
+func newPDServiceAPIClient(client ServiceClient, f errFn) *pdServiceAPIClient {
+	return &pdServiceAPIClient{
+		ServiceClient: client,
+		fn:            f,
+	}
+}
+
+// Available implements ServiceClient.
+func (c *pdServiceAPIClient) Available() bool {
+	return c.ServiceClient.Available() && !c.unavailable.Load()
+}
+
+func (c *pdServiceAPIClient) markAsAvailable() {
+	if !c.unavailable.Load() {
+		return
+	}
+	until := c.unavailableUntil.Load().(time.Time)
+	if time.Now().After(until) {
+		c.unavailable.Store(false)
+	}
+}
+
+// NeedRetry implements ServiceClient.
+func (c *pdServiceAPIClient) NeedRetry(pdErr *pdpb.Error, err error) bool {
+	if c.IsLeader() {
+		return false
+	}
+	if err == nil && pdErr == nil {
+		return false
+	}
+	if c.fn(pdErr) && c.unavailable.CompareAndSwap(false, true) {
+		c.unavailableUntil.Store(time.Now().Add(time.Second * 10))
+		failpoint.Inject("fastCheckAvailable", func() {
+			c.unavailableUntil.Store(time.Now().Add(time.Millisecond * 100))
+		})
+	}
+	return true
+}
+
+// pdServiceBalancerNode is a balancer node for PD service.
+// It extends the pdServiceClient and adds additional fields for the next polling client in the chain.
+type pdServiceBalancerNode struct {
+	ServiceClient
+	next *pdServiceBalancerNode
+}
+
+// pdServiceBalancer is a load balancer for PD service clients.
+// It is used to balance the request to all servers and manage the connections to multiple PD service nodes.
+type pdServiceBalancer struct {
+	mu        sync.Mutex
+	now       *pdServiceBalancerNode
+	totalNode int
+}
+
+func (c *pdServiceBalancer) set(clients []ServiceClient) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if len(clients) == 0 {
+		return
+	}
+	c.totalNode = len(clients)
+	head := &pdServiceBalancerNode{
+		ServiceClient: clients[0],
+	}
+	head.next = head
+	last := head
+	for i := 1; i < c.totalNode; i++ {
+		next := &pdServiceBalancerNode{
+			ServiceClient: clients[i],
+			next:          head,
+		}
+		head = next
+		last.next = head
+	}
+	c.now = head
+}
+
+func (c *pdServiceBalancer) next() {
+	c.now = c.now.next
+}
+
+func (c *pdServiceBalancer) get() (ret ServiceClient) {
+	c.mu.Lock()
+	i := 0
+	defer c.mu.Unlock()
+	if c.now == nil {
+		return nil
+	}
+	for ; i < c.totalNode; i++ {
+		if c.now.Available() {
```

Comment:
If it's not available, will it be excluded in the next round?

Reply:
No. I don't remove unavailable nodes from the queue. The simple approach is taken now and can be optimized in the future if needed